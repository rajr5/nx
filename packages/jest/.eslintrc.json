{
  "extends": "../../.eslintrc",
  "rules": {},
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": ["packages/jest/tsconfig.*?.json"]
      },
      "rules": {}
    },
    {
      "files": ["**/*.ts"],
      "excludedFiles": ["./src/migrations/**"],
      "rules": {
        "no-restricted-imports": ["error", "@nrwl/workspace"]
      }
    }
  ],
  "ignorePatterns": ["!**/*"]
}
