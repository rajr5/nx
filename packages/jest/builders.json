{
  "$schema": "@angular-devkit/architect/src/builders-schema.json",
  "builders": {
    "jest": {
      "class": "./src/builders/jest/jest.builder",
      "schema": "./src/builders/jest/schema.json",
      "description": "Run Jest unit tests"
    }
  }
}
