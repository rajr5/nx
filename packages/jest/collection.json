{
  "name": "Nx Jest",
  "version": "0.1",
  "schematics": {
    "ng-add": {
      "factory": "./src/schematics/ng-add/ng-add",
      "schema": "./src/schematics/ng-add/schema.json",
      "description": "Add Jest configuration to the workspace",
      "hidden": true
    },
    "jest-project": {
      "factory": "./src/schematics/jest-project/jest-project",
      "schema": "./src/schematics/jest-project/schema.json",
      "description": "Add Jest configuration to a project"
    }
  }
}
