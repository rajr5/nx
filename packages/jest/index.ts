export {
  addPropertyToJestConfig,
  removePropertyFromJestConfig,
} from './src/utils/config/update-config';
export {
  jestConfigObjectAst,
  jestConfigObject,
} from './src/utils/config/functions';
