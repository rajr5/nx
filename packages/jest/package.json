{
  "name": "@nrwl/jest",
  "version": "0.0.1",
  "description": "Jest plugin for Nx",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/nrwl/nx.git"
  },
  "keywords": [
    "Monorepo",
    "Angular",
    "React",
    "Web",
    "Node",
    "Nest",
    "Jest",
    "Unit Testing",
    "CLI"
  ],
  "main": "./index.js",
  "typings": "./index.d.ts",
  "author": "Victor Savkin",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/nrwl/nx/issues"
  },
  "homepage": "https://nx.dev",
  "schematics": "./collection.json",
  "builders": "./executors.json",
  "ng-update": {
    "requirements": {},
    "migrations": "./migrations.json"
  },
  "dependencies": {
    "@nrwl/devkit": "*",
    "identity-obj-proxy": "3.0.0",
    "jest-resolve": "^26.6.2",
    "rxjs": "^6.5.4",
    "tslib": "^2.0.0"
  }
}
