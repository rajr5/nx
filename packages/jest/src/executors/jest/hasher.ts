import { Task, Hash, HasherContext } from '@nrwl/devkit';

export default async function run(
  task: Task,
  context: HasherContext
): Promise<Hash> {
  const jestPluginConfig = context.workspaceConfig.pluginsConfig
    ? (context.workspaceConfig.pluginsConfig['@nrwl/jest'] as any)
    : undefined;
  const filter =
    jestPluginConfig && jestPluginConfig.hashingExcludesTestsOfDeps
      ? 'exclude-tests-of-deps'
      : 'all-files';
  return context.hasher.hashTaskWithDepsAndContext(task, filter);
}
