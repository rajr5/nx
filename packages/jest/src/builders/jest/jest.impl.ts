import {
  BuilderContext,
  createBuilder,
  BuilderOutput
} from '@angular-devkit/architect';

import { from, Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import * as path from 'path';
import { JsonObject } from '@angular-devkit/core';

try {
  require('dotenv').config();
} catch (e) {}

import { runCLI } from 'jest';

if (process.env.NODE_ENV == null || process.env.NODE_ENV == undefined) {
  (process.env as any).NODE_ENV = 'test';
}
export interface JestBuilderOptions extends JsonObject {
  codeCoverage?: boolean;
  config?: string;
  jestConfig: string;
  testFile?: string;
  setupFile?: string;
  tsConfig: string;
  bail?: number;
  ci?: boolean;
  color?: boolean;
  clearCache?: boolean;
  findRelatedTests?: string;
  json?: boolean;
  maxWorkers?: number;
  onlyChanged?: boolean;
  outputFile?: string;
  passWithNoTests?: boolean;
  runInBand?: boolean;
  silent?: boolean;
  testNamePattern?: string;
  testPathPattern?: string[];
  colors?: boolean;
  reporters?: string[];
  verbose?: boolean;
  coverage?: boolean;
  coverageReporters?: string;
  coverageDirectory?: string;
  testResultsProcessor?: string;
  updateSnapshot?: boolean;
  useStderr?: boolean;
  watch?: boolean;
  watchAll?: boolean;
  testLocationInResults?: boolean;
}

export default createBuilder<JestBuilderOptions>(run);

function run(
  options: JestBuilderOptions,
  context: BuilderContext
): Observable<BuilderOutput> {
  options.jestConfig = path.resolve(context.workspaceRoot, options.jestConfig);

  const tsJestConfig = {
    tsConfig: path.resolve(context.workspaceRoot, options.tsConfig)
  };

  // TODO: This is hacky, We should probably just configure it in the user's workspace
  // If jest-preset-angular is installed, apply settings
  try {
    require.resolve('jest-preset-angular');
    Object.assign(tsJestConfig, {
      stringifyContentPathRegex: '\\.(html|svg)$',
      astTransformers: ['jest-preset-angular/InlineHtmlStripStylesTransformer']
    });
  } catch (e) {}

  // merge the jestConfig globals with our 'ts-jest' override
  const jestConfig: { globals: any } = require(options.jestConfig);
  const globals = jestConfig.globals || {};
  Object.assign(globals, {
    'ts-jest': {
      ...(globals['ts-jest'] || {}),
      ...tsJestConfig
    }
  });

  const config: any = {
    _: [],
    config: options.config,
    coverage: options.codeCoverage,
    bail: options.bail,
    ci: options.ci,
    color: options.color,
    json: options.json,
    maxWorkers: options.maxWorkers,
    onlyChanged: options.onlyChanged,
    outputFile: options.outputFile,
    passWithNoTests: options.passWithNoTests,
    runInBand: options.runInBand,
    silent: options.silent,
    testLocationInResults: options.testLocationInResults,
    testNamePattern: options.testNamePattern,
    testPathPattern: options.testPathPattern,
    colors: options.colors,
    verbose: options.verbose,
    coverageReporters: options.coverageReporters,
    coverageDirectory: options.coverageDirectory,
    testResultsProcessor: options.testResultsProcessor,
    updateSnapshot: options.updateSnapshot,
    useStderr: options.useStderr,
    watch: options.watch,
    watchAll: options.watchAll,
    globals: JSON.stringify(globals)
  };

  if (options.setupFile) {
    config.setupFilesAfterEnv = [
      path.resolve(context.workspaceRoot, options.setupFile)
    ];
  }

  if (options.testFile) {
    config._.push(options.testFile);
  }

  if (options.findRelatedTests) {
    const parsedTests = options.findRelatedTests.split(',').map(s => s.trim());
    config._.push(...parsedTests);
    config.findRelatedTests = true;
  }

  if (options.clearCache) {
    config.clearCache = true;
  }

  if (options.reporters && options.reporters.length > 0) {
    config.reporters = options.reporters;
  }

  return from(runCLI(config, [options.jestConfig])).pipe(
    map(results => {
      return {
        success: results.results.success
      };
    })
  );
}
