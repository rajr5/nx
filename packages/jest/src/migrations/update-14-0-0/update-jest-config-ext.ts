import {
  formatFiles,
  joinPathFragments,
  logger,
  offsetFromRoot,
  ProjectConfiguration,
  readJson,
  readProjectConfiguration,
  stripIndents,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { jestConfigObject } from '../../utils/config/functions';
import { dirname, extname, join } from 'path';
import {
  removePropertyFromJestConfig,
  addPropertyToJestConfig,
} from '../../utils/config/update-config';
import { JestExecutorOptions } from '../../executors/jest/schema';
import { forEachExecutorOptions } from '@nrwl/workspace/src/utilities/executor-options-utils';

const allowedExt = ['.ts', '.js'];
let isRootPresetUpdated = false;

function updateJestPreset(
  tree: Tree,
  options: JestExecutorOptions,
  projectName: string
) {
  const oldConfig = jestConfigObject(tree, options.jestConfig);
  if (!oldConfig) {
    return;
  }
  // if using the root preset and the root preset was updated to ts file.
  // then update the jest config
  if (isRootPresetUpdated && oldConfig?.preset?.endsWith('jest.preset.js')) {
    removePropertyFromJestConfig(tree, options.jestConfig, 'preset');
    addPropertyToJestConfig(
      tree,
      options.jestConfig,
      'preset',
      joinPathFragments(
        offsetFromRoot(dirname(options.jestConfig)),
        'jest.preset.ts'
      ),
      { valueAsString: false }
    );
  }
}

function updateTsConfig(tree: Tree, tsConfigPath: string) {
  try {
    updateJson(tree, tsConfigPath, (json) => {
      json.exclude = Array.from(
        new Set([...(json.exclude || []), 'jest.config.ts'])
      );
      return json;
    });
  } catch (e) {
    logger.warn(
      stripIndents`Nx Unable to update ${tsConfigPath}. Please manually ignore the jest.config.ts file.`
    );
  }
}

function isJestConfigValid(tree: Tree, options: JestExecutorOptions) {
  const configExt = extname(options.jestConfig);

  if (!tree.exists(options.jestConfig) || !allowedExt.includes(configExt)) {
    logger.debug(
      `unable to update file because it doesn't exist or is not a js or ts file. Config: ${
        options.jestConfig
      }. Exists?: ${tree.exists(options.jestConfig)}`
    );
    return false;
  }
  return true;
}

function updateTsconfigSpec(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  path
) {
  updateJson(tree, joinPathFragments(projectConfig.root, path), (json) => {
    json.include = Array.from(
      new Set([...(json.include || []), 'jest.config.ts'])
    );
    return json;
  });
}

export async function updateJestConfigExt(tree: Tree) {
  if (tree.exists('jest.config.js')) {
    tree.rename('jest.config.js', 'jest.config.ts');
  }

  if (tree.exists('jest.preset.js')) {
    isRootPresetUpdated = true;
    tree.rename('jest.preset.js', 'jest.preset.ts');
  }

  forEachExecutorOptions<JestExecutorOptions>(
    tree,
    '@nrwl/jest:jest',
    (options, projectName, target, configuration) => {
      const projectConfig = readProjectConfiguration(tree, projectName);

      if (!options.jestConfig || !isJestConfigValid(tree, options)) {
        return;
      }

      updateJestPreset(tree, options, projectName);

      const newJestConfigPath = options.jestConfig.replace('.js', '.ts');
      tree.rename(options.jestConfig, newJestConfigPath);

      const rootFiles = tree.children(projectConfig.root);
      for (const fileName of rootFiles) {
        if (fileName === 'tsconfig.json') {
          const filePath = joinPathFragments(projectConfig.root, fileName);
          const tsConfig = readJson(tree, filePath);

          if (tsConfig.references) {
            for (const { path } of tsConfig.references) {
              if (path.endsWith('tsconfig.spec.json')) {
                updateTsconfigSpec(tree, projectConfig, path);
                continue;
              }

              updateTsConfig(tree, joinPathFragments(projectConfig.root, path));
            }
          } else {
            updateTsConfig(tree, filePath);
          }
        }
      }

      projectConfig.targets[target].options.jestConfig = newJestConfigPath;
      updateProjectConfiguration(tree, projectName, projectConfig);
    }
  );
  await formatFiles(tree);
}

export default updateJestConfigExt;
