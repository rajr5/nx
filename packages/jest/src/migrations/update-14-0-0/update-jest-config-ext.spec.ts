import {
  readJson,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { jestInitGenerator } from '@nrwl/jest';
import { libraryGenerator as workspaceLib } from '@nrwl/workspace';
import { updateJestConfigExt } from './update-jest-config-ext';

const setupDefaults = {
  js: true,
  skipPackageJson: true,
  libName: 'lib-one',
  setParserOptionsProject: false,
};

async function libSetUp(tree: Tree, options = setupDefaults) {
  jestInitGenerator(tree, {
    js: options.js,
    skipPackageJson: options.skipPackageJson,
  });
  await workspaceLib(tree, {
    name: options.libName,
    setParserOptionsProject: options.setParserOptionsProject,
  });
  tree.rename(
    `libs/${options.libName}/jest.config.ts`,
    `libs/${options.libName}/jest.config.js`
  );
  const config = tree.read(`libs/${options.libName}/jest.config.js`, 'utf-8');
  tree.write(
    `libs/${options.libName}/jest.config.js`,
    config
      .replace(/\/\* eslint-disable \*\//g, '')
      .replace(/export default/g, 'module.exports =')
  );
  updateProjectConfiguration(tree, options.libName, {
    ...readProjectConfiguration(tree, options.libName),
    targets: {
      test: {
        executor: '@nrwl/jest:jest',
        options: {
          jestConfig: `libs/${options.libName}/jest.config.js`,
          passWithNoTests: true,
        },
        configurations: {
          production: {
            silent: true,
          },
        },
      },
    },
  });
}

describe('Jest Migration (v14.0.0)', () => {
  let tree: Tree;
  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace(2);
  });

  it('should rename project jest.config.js to jest.config.ts', async () => {
    await libSetUp(tree);

    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.read('libs/lib-one/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should rename root jest.config.js', async () => {
    await libSetUp(tree);

    await updateJestConfigExt(tree);
    expect(tree.exists('jest.config.ts')).toBeTruthy();
    expect(tree.exists('jest.preset.js')).toBeTruthy();
  });

  it('should NOT update jest.config.ts preset', async () => {
    await libSetUp(tree);

    tree.rename('libs/lib-one/jest.config.js', 'libs/lib-one/jest.config.ts');
    const projectConfig = readProjectConfiguration(tree, 'lib-one');
    updateProjectConfiguration(tree, 'lib-one', {
      ...projectConfig,
      targets: {
        test: {
          ...projectConfig.targets.test,
          options: {
            jestConfig: 'libs/lib-one/jest.config.ts',
            passWithNoTests: true,
          },
        },
      },
    });
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.read('libs/lib-one/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should only update js/ts files', async () => {
    await libSetUp(tree);

    tree.rename('libs/lib-one/jest.config.js', 'libs/lib-one/jest.config.ts');
    updateProjectConfiguration(tree, 'lib-one', {
      ...readProjectConfiguration(tree, 'lib-one'),
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {
            jestConfig: 'libs/lib-one/jest.config.ts',
            passWithNoTests: true,
          },
        },
      },
    });

    await libSetUp(tree, { ...setupDefaults, libName: 'lib-two' });
    tree.delete('libs/lib-two/jest.config.ts'); // lib generator creates a ts file
    tree.write('libs/lib-two/jest.config.json', '{}');
    updateProjectConfiguration(tree, 'lib-two', {
      ...readProjectConfiguration(tree, 'lib-two'),
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {
            jestConfig: 'libs/lib-two/jest.config.json',
            passWithNoTests: true,
          },
        },
      },
    });

    await libSetUp(tree, { ...setupDefaults, libName: 'lib-three' });
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.exists('libs/lib-two/jest.config.ts')).toBeFalsy();
    expect(tree.exists('libs/lib-two/jest.config.json')).toBeTruthy();
    expect(tree.exists('libs/lib-three/jest.config.ts')).toBeTruthy();
  });

  it('should not throw error if file does not exit', async () => {
    await libSetUp(tree);

    tree.delete('libs/lib-one/jest.config.js');
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeFalsy();
    expect(tree.exists('libs/lib-one/jest.config.js')).toBeFalsy();
  });

  it('should update correct tsconfigs', async () => {
    await libSetUp(tree);

    updateJson(tree, 'libs/lib-one/tsconfig.lib.json', (json) => {
      json.exclude = ['**/*.spec.ts'];
      return json;
    });

    updateJson(tree, 'libs/lib-one/tsconfig.spec.json', (json) => {
      json.include = ['**/*.spec.ts'];
      return json;
    });

    await updateJestConfigExt(tree);

    const tsconfig = readJson(tree, 'libs/lib-one/tsconfig.json');
    const libTsConfig = readJson(tree, 'libs/lib-one/tsconfig.lib.json');
    const specTsConfig = readJson(tree, 'libs/lib-one/tsconfig.spec.json');

    expect(tsconfig.exclude).toBeFalsy();
    expect(libTsConfig.exclude).toEqual(['**/*.spec.ts', 'jest.config.ts']);
    expect(specTsConfig.exclude).toBeFalsy();
    expect(specTsConfig.include).toEqual(['**/*.spec.ts', 'jest.config.ts']);
  });

  it('should add exclude to root tsconfig with no references', async () => {
    await libSetUp(tree);

    tree.delete('libs/lib-one/tsconfig.spec.json');
    tree.delete('libs/lib-one/tsconfig.lib.json');

    updateJson(tree, 'libs/lib-one/tsconfig.json', (json) => {
      delete json.references;
      return json;
    });

    await updateJestConfigExt(tree);

    const tsconfig = readJson(tree, 'libs/lib-one/tsconfig.json');

    expect(tsconfig.exclude).toEqual(['jest.config.ts']);
    expect(tree.exists('libs/lib-one/tsconfig.spec.json')).toBeFalsy();
    expect(tree.exists('libs/lib-one/tsconfig.lib.json')).toBeFalsy();
  });

  it('should update the excludes of next js apps using the project parser settings', async () => {
    await libSetUp(tree, { ...setupDefaults, setParserOptionsProject: true });

    const projectConfig = readProjectConfiguration(tree, 'lib-one');
    projectConfig.targets['build'] = {
      executor: '@nrwl/next:build',
      options: {},
    };
    updateProjectConfiguration(tree, 'lib-one', projectConfig);
    updateJson(tree, 'libs/lib-one/tsconfig.json', (json) => {
      // simulate nextJS tsconfig;
      json.exclude = ['node_modules'];
      return json;
    });
    const esLintJson = readJson(tree, 'libs/lib-one/.eslintrc.json');
    // make sure the parserOptions are set correctly
    expect(esLintJson.overrides[0]).toMatchSnapshot();

    await updateJestConfigExt(tree);

    const tsconfigSpec = readJson(tree, 'libs/lib-one/tsconfig.spec.json');
    expect(tsconfigSpec.exclude).toEqual(['node_modules']);
  });
});
