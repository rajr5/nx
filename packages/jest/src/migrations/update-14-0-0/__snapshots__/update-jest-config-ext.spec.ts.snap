// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Migration (v14.0.0) should rename project jest.config.js to jest.config.ts 1`] = `
"module.exports = {
  displayName: 'lib-one',
  
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    }
  },
  transform: {
    '^.+\\\\\\\\.[tj]sx?$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib-one',\\"preset\\": \\"../../jest.preset.ts\\"
};
"
`;

exports[`Jest Migration (v14.0.0) should update jest.config.ts preset to use the jest.preset.ts 1`] = `
"module.exports = {
  displayName: 'lib-one',
  
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    }
  },
  transform: {
    '^.+\\\\\\\\.[tj]sx?$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib-one',\\"preset\\": \\"../../jest.preset.ts\\"
};
"
`;
